ocular/src/keyring.rs

```diff
@@ -0,0 +1,727 @@
+const COSMOS_ADDRESS_PREFIX: &str = "cosmos";
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum KeyRingType {
+    File,
+    // TODO: OS (and other) Types
+}
+
+impl Default for KeyRingType {
+    fn default() -> Self {
+        KeyRingType::File
+    }
+}
+
+/// Basic keystore traits that all backends are expected to implement
+pub trait KeyStore {
+    /// Create new key store
+    fn create_key_store(&mut self) -> Result<(), KeyStoreError>;
+
+    /// Check if key store has been initialized.
+    fn key_store_created(&self) -> bool;
+
+    /// Check if key exists under specific name. Will return false if no key is found.
+    fn key_exists(&self, name: &str) -> Result<bool, KeyStoreError>;
+
+    /// Add a new key based off of name, password, and derivation path (defaults to cosmos). If override_if_exists is set to true, it will override any existing key with the same name.
+    fn add_key(
+        &self,
+        name: &str,
+        password: &str,
+        derivation_path: Option<&str>,
+        override_if_exists: bool,
+    ) -> Result<PrivateKeyOutput, KeyStoreError>;
+
+    /// Delete key with a given name. If no key exists under name specified an error will be thrown.
+    fn delete_key(&self, name: &str) -> Result<(), KeyStoreError>;
+
+    /// Rename key. If override_if_exists is true any keys with the new key name will be forecfully overriden.
+    fn rename_key(
+        &self,
+        current_name: &str,
+        new_name: &str,
+        override_if_exists: bool,
+    ) -> Result<(), KeyStoreError>;
+
+    /// Get key address in bech32 (aka segwit) format. Will throw an error if the key does not exist.
+    fn get_key_address(&self, name: &str) -> Result<PublicKeyOutput, KeyStoreError>;
+
+    /// Get all key addresses in bech32 (aka segwit) format.
+    fn get_all_keys(&self) -> Result<Vec<PublicKeyOutput>, Box<dyn std::error::Error>>;
+
+    /// Recover key via mnemonic, password, and derivation_path (defaults to cosmos). If override_if_exists is set to true, it will override any existing key with the same name.
+    fn recover_from_mnemonic(
+        &self,
+        name: &str,
+        mnemonic: &str,
+        password: &str,
+        derivation_path: Option<&str>,
+        override_if_exists: bool,
+    ) -> Result<(), KeyStoreError>;
+}
+
+/// Mnemonic and private key in pkcs8 PrivateKeyDocument format
+pub struct PrivateKeyOutput {
+    pub mnemonic: Mnemonic,
+    pub private_key: PrivateKeyDocument,
+}
+
+/// Key name and address in Bech32 (aka segwit) format
+// TODO: Include actual public key here if ever needed
+#[derive(Debug)]
+pub struct PublicKeyOutput {
+    pub name: String,
+    pub account: cosmrs::AccountId,
+}
+
+// --- Base Key Ring ---
+/// Base Keyring that needs to be initialized before being used. Initialization parameters vary depending on type of key store being used.
+pub struct Keyring {
+    pub backend: KeyRingType,
+    pub key_store: Box<dyn KeyStore>,
+}
```

Comment:
One open point of discussion that I haven't 100% figured out yet is how closely we would like to use signatory's keychain and keystore mechanisms. 

As far as requirements go what I've garnered from previous discussions is that we'll likely want to support the file system and potentially some sort of AWS or Azure key store mechanism initially. We additionally want to have the above outlined key access/manipulation methods to support related gravity bridge/steward use cases.

One way of approaching this is to simply make the key_store(s) private variables, add a signatory keyring, and implement the desired traits on top of the signatory keyring depending on the user specified key_store backend. Another approach is to simply rename the overall keyring struct to something else so as not to mislead/misrepresent the functionalities present. There's certainly other ways of designing this as well, it would be great to hear more thoughts/opinions on this @cbrit & @tony-iqlusion/ @tarcieri

Reply:
Signatory's keyrings are designed to support multiple backing keystores, rather than being tied to any particular one (although it only provides one at this time).

They're also designed to accommodate hardware wallets and HSMs.

This is build on the RustCrypto [`signature::Signer`](https://docs.rs/signature/latest/signature/trait.Signer.html) trait as the core abstraction.

> and potentially some sort of AWS or Azure key store mechanism initially

These use cases will most likely require async Rust. There's some initial work to support this sort of thing within the RustCrypto trait ecosystem here: 

https://docs.rs/async-signature/

...however there are no concrete implementations of these traits yet.

Reply:
Cool, as far as overall design then goes, what do you think the best way to represent the signatory keyring in the context of having multiple filestores would be? Should we maintain a reference to a common signatory keyring across file stores within the current 'keyring' struct, or should we do something along the lines of creating a new signatory keyring wrapper struct, and maintain a list of filestore references within that?